server/sonar-server/src/main/java/org/sonar/server/computation/filemove/FileMoveDetectionStep.java

```diff
@@ -0,0 +1,469 @@
+    if (scoreMatrix.maxScore < MIN_REQUIRED_SCORE) {
+      LOG.debug("max score in matrix is less than min required score (%s). Do nothing.", MIN_REQUIRED_SCORE);
+      return;
+    }
+
+    MatchesByScore matchesByScore = MatchesByScore.create(scoreMatrix);
+
+    ElectedMatches electedMatches = electMatches(dbFileSourcesByKey, reportFileSourcesByKey, matchesByScore);
+
+    for (Match validatedMatch : electedMatches) {
+      movedFilesRepository.setOriginalFile(
+        reportFilesByKey.get(validatedMatch.getReportKey()),
+        toOriginalFile(dbFilesByKey.get(validatedMatch.getDbKey())));
+      LOG.info("File move found: " + validatedMatch);
+    }
+  }
+
+  private Map<String, ComponentDtoWithSnapshotId> getDbFilesByKey(Snapshot baseProjectSnapshot) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      return from(dbClient.componentDao().selectAllChildren(
+        dbSession,
+        ComponentTreeQuery.builder()
+          .setBaseSnapshot(new SnapshotDto()
+            .setId(baseProjectSnapshot.getId())
+            .setRootId(baseProjectSnapshot.getId()))
+          .setQualifiers(FILE_QUALIFIERS)
+          .setSortFields(SORT_FIELDS)
+          .setPageSize(Integer.MAX_VALUE)
+          .setPage(1)
+          .build()))
+            .uniqueIndex(ComponentDtoFunctions.toKey());
+    }
+  }
+
+  private static Map<String, Component> getReportFilesByKey(Component root) {
+    final ImmutableMap.Builder<String, Component> builder = ImmutableMap.builder();
+    new DepthTraversalTypeAwareCrawler(
+      new TypeAwareVisitorAdapter(CrawlerDepthLimit.FILE, POST_ORDER) {
+        @Override
+        public void visitFile(Component file) {
+          builder.put(file.getKey(), file);
+        }
+      }).visit(root);
+    return builder.build();
+  }
+
+  private Map<String, File> getDbFileSourcesByKey(Map<String, ComponentDtoWithSnapshotId> dbFilesByKey, Set<String> removedFileKeys) {
+    try (DbSession dbSession = dbClient.openSession(false)) {
+      ImmutableMap.Builder<String, File> builder = ImmutableMap.builder();
+      for (String fileKey : removedFileKeys) {
+        ComponentDtoWithSnapshotId componentDto = dbFilesByKey.get(fileKey);
+        FileSourceDto fileSourceDto = dbClient.fileSourceDao().selectSourceByFileUuid(dbSession, componentDto.uuid());
+        if (fileSourceDto != null) {
+          builder.put(fileKey, new File(componentDto.path(), fileSourceDto.getSrcHash(), on('\n').splitToList(fileSourceDto.getLineHashes())));
+        }
+      }
+      return builder.build();
+    }
+  }
+
+  private Map<String, File> getReportFileSourcesByKey(Map<String, Component> reportFilesByKey, Set<String> addedFileKeys) {
+    ImmutableMap.Builder<String, File> builder = ImmutableMap.builder();
+    for (String fileKey : addedFileKeys) {
+      // FIXME computation of sourceHash and lineHashes might be done multiple times for some files: here, in ComputeFileSourceData, in
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Take the required action to fix the issue indicated by this comment. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS1134)